File: merge_sort/c/merge_sort.c
#include <stdlib.h>
#include "merge_sort.h"

#define array_size(x) ((sizeof(x)/sizeof(0[x])) / ((size_t)(!(sizeof(x) % sizeof(0[x])))))
#define min(a, b) ((a) < (b) ? (a) : (b))

static void merge(int *arr, int *copy, int lo, int mid, int hi)
{
  /* Arr[lo] to arr[mid] and arr[mid + 1] to arr[hi] must be sorted. */
  
  for (int k = lo; k <= hi; k++)
    copy[k] = arr[k];

  int i = lo, j = mid + 1;
  for (int k = lo; k <= hi; k++) {
    if (i > mid)                  arr[k] = copy[j++];
    else if (j > hi)              arr[k] = copy[i++];
    else if (copy[i] < copy[j])   arr[k] = copy[i++];
    else if (copy[i] >= copy[j])  arr[k] = copy[j++];
  }
}

static void sort_(int *arr, int *copy, int lo, int hi)
{
  if (hi <= lo) return;
  int mid = lo + (hi - lo) / 2;
  sort_(arr, copy, lo, mid);
  sort_(arr, copy, mid + 1, hi);
  merge(arr, copy, lo, mid, hi);
}

void sort_iter(int *arr, int size)
{
  int copy[size];
  for (int sz = 1; sz < size; sz = sz + sz)
    for (int lo = 0; lo < size - sz; lo += sz + sz)
      merge(arr, copy, lo, lo + sz - 1, min(lo + sz + sz - 1, size - 1));
}

void sort(int *arr, int size)
{
  int copy[size];
  sort_(arr, copy, 0, size - 1);
}

File: merge_sort/python/merge_sort.py
#!/usr/bin/env python3
from typing import List, Any


def _merge(arr: List[Any], copy: List[Any], lo: int, mid: int,
           hi: int) -> None:
  # Arr[lo] to arr[mid] and arr[mid + 1] to arr[hi] must be sorted.
  copy[lo:hi + 1] = arr[lo:hi + 1]

  i, j = lo, mid + 1
  for k in range(lo, hi + 1):
    if i > mid:
      arr[k] = copy[j]; j += 1
    elif j > hi:
      arr[k] = copy[i]; i += 1
    elif copy[i] < copy[j]:
      arr[k] = copy[i]; i += 1
    elif copy[i] >= copy[j]:
      arr[k] = copy[j]; j += 1

def _sort(arr: List[Any], copy: List[Any], lo: int, hi: int) -> None:
  if hi <= lo:
    return
  mid = lo + (hi - lo) // 2
  _sort(arr, copy, lo, mid)
  _sort(arr, copy, mid + 1, hi)
  _merge(arr, copy, lo, mid, hi)

def sort(arr: List[Any]) -> None:
  copy = arr[:]
  _sort(arr, copy, 0, len(arr) - 1)
